### ISSUE TITLE
Rethink `join` column deduplication

### ISSUE THREAD

**Issue by jcrist** [ibis-project/ibis#5537]
Based on comments in #5514 (and some internal discussions), I think we might want to change how we handle duplicate column names in joins.

## The problem

Given tables `t1` and `t2` like:

```python
import ibis

t1 = ibis.table({"id": "int", "a": "string", "b": "string"})
t2 = ibis.table({"id": "int", "a": "string", "c": "string"})
t3 = ibis.table({"id": "int", "d"})
```

joining these tables on `id` will result in a table with duplicate column names (`a`, and `id` if not an inner join). Ibis requires tables have a valid schema with unique column names after every operation. As such, we can't have two `a` columns in the output, so we need to deduplicate these names. The current strategy is:

- Duplicate column names have suffixes appended to both the left & right versions (defaulting to `"_x"` and `"_y"`)

- *Except* for duplicate columns that are the join keys in an inner join only. These columns are known to be exactly identical (due to the inner join); only one copy is left in the output.

```python
In [4]: t1.join(t2, "id").columns  # inner join, only one copy of `id` kept
Out[4]: ['id', 'a_x', 'b', 'a_y', 'c']

In [5]: t1.left_join(t2, "id").columns  # left join, both `id` and `a` are deduplicated
Out[5]: ['id_x', 'a_x', 'b', 'id_y', 'a_y', 'c']
```

While simple and correct, this behavior can be unergonomic/annoying for a few reasons:

- Chaining multiple non-inner joins where all keys have the same name (e.g. `t1.left_join(t2, "id").left_join(t3, "id")`). In this case the result of the first join will have `id` deduplicated into `id_x` and `id_y`. This causes the second join to error since there's no `id` column anymore.

- The magical `_x` and `_y` suffixes only show up for conflicting column names. When joining large tables it can be hard to know if a conflict will happen, and if so what columns will conflict. Is `name` ok? Or do I need to use `name_x`?

- The suffixes often result in weird names. Every row in a table is generally an entity of some form, it often reads clearer to me to use a prefix instead. For example, in a table of customers, it's clearer to have `customer_name`/`customer_id` rather than `name_customer`/`id_customer`. Both are better than the default of `name_x`/`id_x`.

- Our current mixed support for passing in expressions that don't share a common parent (e.g. `t1.join(t2, "id").join(t3, "id").select(t1.a)`) can lead to invalid or weird queries. In the past we had talked about fully outlawing these cases and erroring nicely. I think this would be good to do, and would eliminate some of the issues here.

Often when using our current API I find myself doing one of two things:

- Passing in `suffixes=("", "_some_name")` to only append a suffix to duplicate columns in the right table, leaving the left columns as they are.

- Renaming all columns in the right table with a prefix before joining using `Table.relabel`. Something like `right.relabel(lambda c: f"prefix_{c}")`.

There are some rough edges here, and I think they can be improved.

**Comment by jcrist** [comment]
After talking with @gforsyth and @kszucs today about this, I have the following loose proposal:

- Deprecate/remove `suffixes` in favor of separate kwargs for `left` and `right`. This is both to provide a clean break in the API, as well as to make it clearer/easier to rename only one of the tables on conflicts. I also think we should make it easy to use a prefix instead of a suffix, as that generally reads better to me. Perhaps `right_prefix`/`left_prefix`? Could also do something more general like `relabel_left`/`relabel_right` that takes either a format string (`"foo_%s"`) or general callable and returns a new name for the conflicting column.

- Default to only renaming duplicate columns in the right table. This makes it easier to join on new columns to an existing table without having to worry about renaming existing columns. The default might be `right_prefix="right_"` or something.

- OR maybe not make this choice for the user at all, and instead error nicely by default if there's a conflict pointing the user to these kwargs. This forces the user to acknowledge the conflict and think about how they want to address it, but would require the user to manually set these kwargs more frequently.

With these changes, the join above might be:

```python
query = (
    t1
    .left_join(t2, "id", right_prefix="t2_")
    .left_join(t3, "id")
    .select("id", "a", "t2_a", "c", "d")
) # id and a are from `t1`, `t2_a` is from t2
```

**Comment by judahrand** [comment]
> Our current mixed support for passing in expressions that don't share a common parent (e.g. t1.join(t2, "id").join(t3, "id").select(t1.a)) can lead to invalid or weird queries. In the past we had talked about fully outlawing these cases and erroring nicely. I think this would be good to do, and would eliminate some of the issues here.

I feel that losing this would be unfortunate. For me at least it is preferable to reference the columns from their parent tables in a final `select` after chained `joins` so that it is clear where the columns came from. This is also most relevant when the tables are large and working out what comes from where is non-obvious.

**Comment by NickCrews** [comment]
> Passing in suffixes=("", "_some_name") to only append a suffix to duplicate columns in the right table, leaving the left columns as they are.

This is what I most often do. Left join a "query" table with a "lookup" table.

I actually prefer suffixes over prefixes, so that if I sort columns then "customer_x" is next to "customer_y". But I could see how prefixes could also be preferred. I like "foo_%s", it is immediately clear what is happening and quite concise. Accepting a lambda would be great too. Feature request: make .relabel() able to take format strings and lambdas!

> OR maybe not make this choice for the user at all, and instead error nicely by default if there's a conflict pointing the user to these kwargs.

The user is almost always going to be working with tables where the schema of those tables is known beforehand (I think). They NEED to be thinking about name conflicts. With the default behavior of adding _x and _y, I've not realized I had column collisions, and have accidentally generated _x and _y suffixed columns that broke later processing steps. It would have been better for me if ibis had just errored.

I think I would prefer it if
- Ibis just errored on column collision
- it was easier for users to relabel columns beforehand
then Ibis doesn't have to worry about this problem, and I think it's less ambiguous for users:

```python
collisions = right.schema() - left.schema()  # feat: request allow set difference for schemas
# union = left.schema() | right.schema()  # not needed here, but relatedly would be nice
right = right.relabel({c: "customer_%s" for c in collisions})
joined = left.left_join(right, "id")
```

**Comment by judahrand** [comment]
> They NEED to be thinking about name conflicts.

I massively agree with this. Especially, given that the schema of the tables being queried can change. I'd argue that even when there aren't name conflicts the user needs to be thinking about possible future conflicts (this is one of the reasons [this](https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.rules.Rule_L027) SQLFluff linting rule is so important)! The silent renaming of columns could easily break code which was previously working because the schema of one of the tables changes. This would then be very difficult to debug. If Ibis errored out nicely and helpfully as soon as a column collision was detected it would make this infinitely easier to debug.

On top of this being able to reference columns in chained joins by their parent tables would make Ibis code much more robust to schema changes. Imagine this example:

```python
import ibis
from ibis import _

t1 = ibis.table({"id": "int", "a": "string"})
t2 = ibis.table({"id": "int", "b": "string"})
t3 = ibis.table({"id": "int", "c": "string"})

(
    t1
    .join(t2, "id")
    .join(t3, "id")
    .select(_.id, _.a, _.b, _.c)
```

Now imagine that `t3` is updated to have the following schema `t3 = ibis.table({"id": "int", "a": "string", "c": "string"})`. The code above will break. If it was instead written like this

```python
import ibis
from ibis import _

t1 = ibis.table({"id": "int", "a": "string"})
t2 = ibis.table({"id": "int", "b": "string"})
t3 = ibis.table({"id": "int", "c": "string"})

(
    t1
    .join(t2, "id")
    .join(t3, "id")
    .select(t1.id, t1.a, t2.b, t3.c)
```

it would be resilient to this schema change. Furthermore, it is much more readable and clear where the output columns come from which aides debugability massively.

**Comment by gforsyth** [comment]
> I feel that losing this would be unfortunate. For me at least it is preferable to reference the columns from their parent tables in a final `select` after chained `joins` so that it is clear where the columns came from. This is also most relevant when the tables are large and working out what comes from where is non-obvious.

We agree that this is a useful feature, but as it exists right now, it's only functional by accident and _usually_ returns the wrong result.  We're thinking about how to add in proper support for this in light of the eager evaluation of the column names.

I suspect we'll first arrive at a place where we've made that sort of usage error, in the interest of not returning incorrect results, but in the longer term we'll hopefully support this type of reference in post-join projections.

**Comment by jcrist** [comment]
> Feature request: make .relabel() able to take format strings and lambdas!

Just to note - `t.relabel` already does support lambdas. The function takes in an old name, and returns a new name or `None` for "no change".

**Comment by jcrist** [comment]
This has been fixed in #5883, and will be part of the 6.0 release. Closing.
